File: 종만북/getPartialMatch.cpp
/***********************************
    
    KMP algorithm

    Input : pattern

    Output : suffix, preffix length array

    Calculates suffix preffix length in pattern


    
    8/21 coded by bmkim
    
***********************************/
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;
vector<int> getPartialMatch(string& s){
    int k = s.size();
    vector<int> ret(k, 0);
    int begin = 1, length = 0;
    while(begin < k){
        if(s[begin] == s[length]){
            length++;
            ret[begin] = length;
            begin++;
        }
        else{
            if(length == 0){
                begin++;
            }
            else{
                length = ret[length-1];
            }
        }
    }
    return ret;
}
vector<int> getPrefixSuffix(string& s){
    vector<int> part = getPartialMatch(s);
    vector<int> ret;
    int length = s.size();
    while(length > 0){
        ret.push_back(length);
        length = part[length-1];
    }
    return ret;
}
int main(){
    string a,b;
    cin >> a >> b;
    a += b;
    vector<int> ret = getPrefixSuffix(a);
    int n = ret.size() - 1;
    while(n >= 0){
        cout << ret[n] << " ";
        n--;
    }
}

File: 종만북/getPartialMatch.py
"""
    
    KMP algorithm

    Input : pattern

    Output : suffix, preffix length array

    Calculates suffix preffix length in pattern


    
    8/16 coded by bmkim
    
"""
def getPartialMatch(word):
    pi = [0] * len(word)
    
    leng = 0
    begin = 1
    while begin < len(word):
        if word[begin] == word[leng]:
            leng += 1
            pi[begin] = leng
            begin += 1
        else:
            if leng != 0:
                leng = pi[leng-1]
            else:
                begin += 1
    return pi


if __name__ == "__main__":
    s = input()
    s += input()
    ret = []
    length = getPartialMatch(s)
    k = len(length)
    while k > 0:
        ret.append(k)
        k = length[k-1]
    n = len(ret) - 1
    while n >= 0:
        print(ret[n], end=" ")
        n -= 1